services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -504,11 +504,6 @@ public static void dettachClientSession(UserSessionProvider sessions, RealmModel
         }
 
         clientSession.detachFromUserSession();
-
-        // TODO: Might need optimization to prevent loading client sessions from cache in getAuthenticatedClientSessions()
-        if (userSession.getAuthenticatedClientSessions().isEmpty()) {
```

Comment:
Wouldn't that impact client code relying on this method to also remove user sessions? For instance, token revocation and, potentially, logout?

I run a quick test `org.keycloak.testsuite.oauth.TokenRevocationTest#testRevokeToken` after adding the following assertions at the end of the method:

```
response = oauth.doTokenRevoke(tokenResponse2.getRefreshToken(), "refresh_token", "password");
assertThat(response, Matchers.statusCodeIsHC(Status.OK));

userSessions = testUser.getUserSessions();
assertEquals(0, userSessions.size());
```

The code above should force user session invalidation because all client sessions were revoked. Currently, it expects no user sessions after revoking the last client session. With your changes, the user session is still active.

Reply:
@pedroigor (cc @stianst) Yes, that is exactly the point of the PR to not remove userSession in case that clientSession is removed due the consent revoke or token revoke.

In current master, SSO session is entirely removed when consent (or token) is revoked, in case that there is no other client authenticated to SSO session. IMO this is not correct behaviour and it is exactly the reason of this bug scenario described in the https://issues.redhat.com/browse/KEYCLOAK-16006 . In my view, revoke single consent (or token) is a thing specific to single client, not to whole SSO session.

For the logout, the SSO session is still removed together with all clientSessions attached to the userSession. This is not changed in this PR.

Reply:
I understand that. But considering token revocation does not make sense to keep the user session open if there is no client session associated with the user session.

Reply:
@pedroigor hmm... I am not 100% sure about that TBH. My vote is that token revocation should not invalidate whole SSO session. If the session is leaked, user should not invalidate single token, but rather should logout from his sessions. Which he can do with "Logout" or through account management....

However since this bug is about consent, I can change it in a way that revoke of the token will still invalidate SSO session in case no more clients. But revoke consent will still keep SSO session open. I can include the test snippet you added in previous comment in the PR to make sure that token revocation invalidates user session. Will this work for you?

Reply:
Sure. So we keep the current behavior.

Reply:
Just to be clear, you want to keep the current behaviour for the token revocation, but you are fine to change it for the consent  to fix the issue described in the steps to reproduce of https://issues.redhat.com/browse/KEYCLOAK-16006 . Is it correct?

Reply:
Yes.

Reply:
@pedroigor As discussed, I've updated the PR to remove userSession after revoke last clientSession. Do you please have a chance to re-review?